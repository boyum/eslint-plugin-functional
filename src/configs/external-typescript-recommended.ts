import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import externalVanillaRecommended from "#eslint-plugin-functional/configs/external-vanilla-recommended";

const tsConfig = {
  "@typescript-eslint/prefer-readonly": "error",
  "@typescript-eslint/switch-exhaustiveness-check": "error",
} satisfies FlatConfig.Config["rules"];

export default {
  ...externalVanillaRecommended,
  ...tsConfig,
} satisfies FlatConfig.Config["rules"];
