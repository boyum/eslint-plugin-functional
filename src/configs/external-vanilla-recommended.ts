import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

export default {
  "prefer-const": "error",
  "no-param-reassign": "error",
  "no-var": "error",
} satisfies FlatConfig.Config["rules"];
