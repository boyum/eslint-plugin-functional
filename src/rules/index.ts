import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClassInheritance from "./no-class-inheritance";
import * as noClasses from "./no-classes";
import * as noConditionalStatements from "./no-conditional-statements";
import * as noExpressionStatements from "./no-expression-statements";
import * as noLet from "./no-let";
import * as noLoopStatements from "./no-loop-statements";
import * as noMixedTypes from "./no-mixed-types";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpressions from "./no-this-expressions";
import * as noThrowStatements from "./no-throw-statements";
import * as noTryStatements from "./no-try-statements";
import * as preferImmutableTypes from "./prefer-immutable-types";
import * as preferPropertySignatures from "./prefer-property-signatures";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferTacit from "./prefer-tacit";
import * as readonlyType from "./readonly-type";
import * as typeDeclarationImmutability from "./type-declaration-immutability";

/**
 * All of the custom rules.
 */
export const rules: Readonly<{
  [functionalParameters.name]: typeof functionalParameters.rule;
  [immutableData.name]: typeof immutableData.rule;
  [noClasses.name]: typeof noClasses.rule;
  [noClassInheritance.name]: typeof noClassInheritance.rule;
  [noConditionalStatements.name]: typeof noConditionalStatements.rule;
  [noExpressionStatements.name]: typeof noExpressionStatements.rule;
  [noLet.name]: typeof noLet.rule;
  [noLoopStatements.name]: typeof noLoopStatements.rule;
  [noMixedTypes.name]: typeof noMixedTypes.rule;
  [noPromiseReject.name]: typeof noPromiseReject.rule;
  [noReturnVoid.name]: typeof noReturnVoid.rule;
  [noThisExpressions.name]: typeof noThisExpressions.rule;
  [noThrowStatements.name]: typeof noThrowStatements.rule;
  [noTryStatements.name]: typeof noTryStatements.rule;
  [preferImmutableTypes.name]: typeof preferImmutableTypes.rule;
  [preferPropertySignatures.name]: typeof preferPropertySignatures.rule;
  [preferReadonlyTypes.name]: typeof preferReadonlyTypes.rule;
  [preferTacit.name]: typeof preferTacit.rule;
  [readonlyType.name]: typeof readonlyType.rule;
  [typeDeclarationImmutability.name]: typeof typeDeclarationImmutability.rule;
}> = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClasses.name]: noClasses.rule,
  [noClassInheritance.name]: noClassInheritance.rule,
  [noConditionalStatements.name]: noConditionalStatements.rule,
  [noExpressionStatements.name]: noExpressionStatements.rule,
  [noLet.name]: noLet.rule,
  [noLoopStatements.name]: noLoopStatements.rule,
  [noMixedTypes.name]: noMixedTypes.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpressions.name]: noThisExpressions.rule,
  [noThrowStatements.name]: noThrowStatements.rule,
  [noTryStatements.name]: noTryStatements.rule,
  [preferImmutableTypes.name]: preferImmutableTypes.rule,
  [preferPropertySignatures.name]: preferPropertySignatures.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferTacit.name]: preferTacit.rule,
  [readonlyType.name]: readonlyType.rule,
  [typeDeclarationImmutability.name]: typeDeclarationImmutability.rule,
};
