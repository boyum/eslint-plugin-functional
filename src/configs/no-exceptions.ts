import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#eslint-plugin-functional/rules";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

export default Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.category === "No Exceptions" &&
        rule.meta.docs.recommended !== false,
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];
