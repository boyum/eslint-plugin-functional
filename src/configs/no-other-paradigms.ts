import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#eslint-plugin-functional/rules";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

export default Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.category === "No Other Paradigms" &&
        rule.meta.docs.recommended !== false,
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];
