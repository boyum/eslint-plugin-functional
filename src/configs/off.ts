import type { FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#/rules";
import { ruleNameScope } from "#/utils/misc";

export default Object.fromEntries(
  Object.entries(rules).map(([name]) => [`${ruleNameScope}/${name}`, "off"]),
) satisfies FlatConfig.Config["rules"] as NonNullable<FlatConfig.Config["rules"]>;
