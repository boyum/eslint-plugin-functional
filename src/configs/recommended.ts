import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";
import { Immutability } from "is-immutable-type";

import { rules } from "#eslint-plugin-functional/rules";
import * as functionalParameters from "#eslint-plugin-functional/rules/functional-parameters";
import * as noConditionalStatements from "#eslint-plugin-functional/rules/no-conditional-statements";
import * as noLet from "#eslint-plugin-functional/rules/no-let";
import * as noThisExpressions from "#eslint-plugin-functional/rules/no-this-expressions";
import * as noThrowStatements from "#eslint-plugin-functional/rules/no-throw-statements";
import * as noTryStatements from "#eslint-plugin-functional/rules/no-try-statements";
import * as preferImmutableTypes from "#eslint-plugin-functional/rules/prefer-immutable-types";
import * as typeDeclarationImmutability from "#eslint-plugin-functional/rules/type-declaration-immutability";
import { RuleEnforcementComparator } from "#eslint-plugin-functional/rules/type-declaration-immutability";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

const recommended = Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.recommended === "recommended" &&
        rule.meta.docs.category !== "Stylistic",
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];

const overrides = {
  [functionalParameters.fullName]: [
    "error",
    {
      enforceParameterCount: {
        ignoreLambdaExpression: true,
        ignoreIIFE: true,
        ignoreGettersAndSetters: true,
      },
    },
  ],
  [noConditionalStatements.fullName]: [
    "error",
    {
      allowReturningBranches: true,
    },
  ],
  [noLet.fullName]: [
    "error",
    {
      allowInForLoopInit: true,
    },
  ],
  [noThisExpressions.fullName]: "off",
  [noThrowStatements.fullName]: [
    "error",
    {
      allowInAsyncFunctions: true,
    },
  ],
  [noTryStatements.fullName]: "off",
  [preferImmutableTypes.fullName]: [
    "error",
    {
      enforcement: "None",
      ignoreInferredTypes: true,
      parameters: {
        enforcement: "ReadonlyDeep",
      },
    },
  ],
  [typeDeclarationImmutability.fullName]: [
    "error",
    {
      rules: [
        {
          identifiers: ["^I?Immutable.+"],
          immutability: Immutability.Immutable,
          comparator: RuleEnforcementComparator.AtLeast,
        },
        {
          identifiers: ["^I?ReadonlyDeep.+"],
          immutability: Immutability.ReadonlyDeep,
          comparator: RuleEnforcementComparator.AtLeast,
        },
        {
          identifiers: ["^I?Readonly.+"],
          immutability: Immutability.ReadonlyShallow,
          comparator: RuleEnforcementComparator.AtLeast,
          fixer: [
            {
              pattern: "^(Array|Map|Set)<(.+)>$",
              replace: "Readonly$1<$2>",
            },
            {
              pattern: "^(.+)$",
              replace: "Readonly<$1>",
            },
          ],
        },
        {
          identifiers: ["^I?Mutable.+"],
          immutability: Immutability.Mutable,
          comparator: RuleEnforcementComparator.AtMost,
          fixer: [
            {
              pattern: "^Readonly(Array|Map|Set)<(.+)>$",
              replace: "$1<$2>",
            },
            {
              pattern: "^Readonly<(.+)>$",
              replace: "$1",
            },
          ],
        },
      ],
    },
  ],
} satisfies FlatConfig.Config["rules"];

export default {
  ...recommended,
  ...overrides,
};
