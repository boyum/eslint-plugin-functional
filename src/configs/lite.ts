import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import * as functionalParameters from "#eslint-plugin-functional/rules/functional-parameters";
import * as immutableData from "#eslint-plugin-functional/rules/immutable-data";
import * as noConditionalStatements from "#eslint-plugin-functional/rules/no-conditional-statements";
import * as noExpressionStatements from "#eslint-plugin-functional/rules/no-expression-statements";
import * as preferImmutableTypes from "#eslint-plugin-functional/rules/prefer-immutable-types";

import recommended from "./recommended";

const overrides = {
  [functionalParameters.fullName]: [
    "error",
    {
      enforceParameterCount: false,
    },
  ],
  [immutableData.fullName]: ["error", { ignoreClasses: "fieldsOnly" }],
  [noConditionalStatements.fullName]: "off",
  [noExpressionStatements.fullName]: "off",
  [preferImmutableTypes.fullName]: [
    "error",
    {
      enforcement: "None",
      ignoreInferredTypes: true,
      parameters: {
        enforcement: "ReadonlyShallow",
      },
    },
  ],
} satisfies FlatConfig.Config["rules"];

export default {
  ...recommended,
  ...overrides,
};
