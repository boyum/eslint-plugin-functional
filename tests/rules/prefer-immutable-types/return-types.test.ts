import dedent from "dedent";
import { createRuleTester } from "eslint-vitest-rule-tester";
import { describe, expect, it } from "vitest";

import { name, rule } from "#/rules/prefer-immutable-types";

import { typescriptConfig } from "../../utils/configs";

describe(name, () => {
  describe("typescript", () => {
    const { valid, invalid } = createRuleTester({
      name,
      rule,
      configs: typescriptConfig,
    });

    it("reports non-immutable set return types", () => {
      const invalidResult = invalid({
        code: "function foo(): ReadonlySet<string> {}",
        errors: ["returnType"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports non-immutable map return types", () => {
      const invalidResult = invalid({
        code: "function foo(): ReadonlyMap<string, string> {}",
        errors: ["returnType"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports mutable records return types", () => {
      const invalidResult = invalid({
        code: "function foo(): { foo: string } {}",
        errors: ["returnType"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports mutable records return types and suggests a fix for ReadonlyShallow", () => {
      const invalidResult = invalid({
        code: "function foo(): { foo: string } {}",
        options: [{ returnTypes: "ReadonlyShallow" }],
        errors: ["returnType"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
      for (const message of invalidResult.messages) {
        expect(message.suggestions).toHaveLength(1);
      }
    });

    it("doesn't report type assertion return types", () => {
      valid({
        code: "function foo(arg0: { foo: string | number }, arg1: { foo: string | number }): arg0 is { foo: number } {}",
        options: [
          {
            parameters: false,
            returnTypes: "ReadonlyShallow",
          },
        ],
      });
    });

    it("suggest multiple fixes for collections for ReadonlyShallow", () => {
      const invalidResult = invalid({
        code: dedent`
          function foo(): Array<string> {}
          function foo(): string[] {}
          function foo(): Set<string> {}
          function foo(): Map<string, string> {}
          function foo(): ReadonlyArray<string> {}
          function foo(): readonly string[] {}
          function foo(): ReadonlySet<string> {}
          function foo(): ReadonlyMap<string, string> {}
        `,
        options: [{ returnTypes: "ReadonlyShallow" }],
        errors: ["returnType", "returnType", "returnType", "returnType"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
      for (const message of invalidResult.messages) {
        expect(message.suggestions).toHaveLength(2);
      }
    });

    it("allows for user suggestions", () => {
      const invalidResult = invalid({
        code: "function foo(): { foo: string } {}",
        options: [
          {
            returnTypes: "ReadonlyDeep",
            suggestions: {
              ReadonlyDeep: [
                [
                  {
                    pattern: "^(?!ReadonlyDeep)(?:Readonly<(.+)>|(.+))$",
                    replace: "ReadonlyDeep<$1$2>",
                  },
                ],
              ],
            },
          },
        ],
        errors: ["returnType"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
      for (const message of invalidResult.messages) {
        expect(message.suggestions).toHaveLength(1);
      }
    });

    it("allows for user fixes", () => {
      const invalidResult = invalid({
        code: dedent`
          function foo(): Array<string> {}
          function foo(): string[] {}
          function foo(): Set<string> {}
          function foo(): Map<string, string> {}
          function foo(): ReadonlyArray<string> {}
          function foo(): readonly string[] {}
          function foo(): ReadonlySet<string> {}
          function foo(): ReadonlyMap<string, string> {}
        `,
        options: [
          {
            returnTypes: "Immutable",
            fixer: {
              Immutable: [
                {
                  pattern: "^(?:Readonly)?(Set|Map)<(.+)>$",
                  replace: "Readonly<Readonly$1<$2>>",
                },
              ],
            },
          },
        ],
        errors: [
          "returnType",
          "returnType",
          "returnType",
          "returnType",
          "returnType",
          "returnType",
          "returnType",
          "returnType",
        ],
        verifyAfterFix: false, // "fix" doesn't fix arrays so they will still report.
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it.each([
      [[{ returnTypes: "ReadonlyShallow" }]],
      [[{ returnTypes: "ReadonlyDeep" }]],
      [[{ returnTypes: "Immutable" }]],
    ])("doesn't reports valid return types", (options) => {
      valid({
        code: "function foo(): boolean {}",
        options,
      });
      valid({
        code: "function foo(): true {}",
        options,
      });
      valid({
        code: "function foo(): string {}",
        options,
      });
      valid({
        code: "function foo(): 'bar' {}",
        options,
      });
      valid({
        code: "function foo(): undefined {}",
        options,
      });
      valid({
        code: "function foo(): readonly string[] {}",
        options,
        settings: {
          immutability: {
            overrides: [
              {
                type: { from: "lib", name: "ReadonlyArray" },
                to: "Immutable",
              },
            ],
          },
        },
      });
      valid({
        code: "function foo(): { readonly foo: string } {}",
        options,
      });
      valid({
        code: "function foo(): { readonly foo: { readonly bar: number } } {}",
        options,
      });
      valid({
        code: "function foo(): Readonly<ReadonlySet<string>> {}",
        options,
      });
      valid({
        code: "function foo(): Readonly<ReadonlyMap<string, string>> {}",
        options,
      });
      valid({
        code: "function foo(arg: { foo: string | number }): arg is { foo: number } {}",
        options,
      });

      if (options[0]!.returnTypes !== "Immutable") {
        valid({
          code: "function foo(): { foo(): void } {}",
          options,
        });
        valid({
          code: "function foo(): ReadonlyArray<string> {}",
          options,
        });
        valid({
          code: "function foo(): readonly [string, number] {}",
          options,
        });
        valid({
          code: "function foo(): Readonly<[string, number]> {}",
          options,
        });
        valid({
          code: "function foo(): { foo: () => void } {}",
          options,
        });
        valid({
          code: "function foo(): ReadonlySet<string> {}",
          options,
        });
        valid({
          code: "function foo(): ReadonlyMap<string, string> {}",
          options,
        });
        valid({
          code: dedent`
            interface Foo {
              (): readonly string[];
            }
          `,
          options,
        });
        valid({
          code: dedent`
            interface Foo {
              new (): readonly string[];
            }
          `,
          options,
        });
      }
    });
  });
});
