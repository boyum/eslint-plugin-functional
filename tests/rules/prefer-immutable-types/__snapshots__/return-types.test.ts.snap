// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefer-immutable-types > typescript > allows for user fixes 1`] = `
[
  {
    "column": 15,
    "endColumn": 30,
    "endLine": 1,
    "line": 1,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
  {
    "column": 15,
    "endColumn": 25,
    "endLine": 2,
    "line": 2,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
  {
    "column": 15,
    "endColumn": 28,
    "endLine": 3,
    "fix": {
      "range": [
        77,
        88,
      ],
      "text": "Readonly<ReadonlySet<string>>",
    },
    "line": 3,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
  {
    "column": 15,
    "endColumn": 36,
    "endLine": 4,
    "fix": {
      "range": [
        108,
        127,
      ],
      "text": "Readonly<ReadonlyMap<string, string>>",
    },
    "line": 4,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
  {
    "column": 15,
    "endColumn": 38,
    "endLine": 5,
    "line": 5,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
  {
    "column": 15,
    "endColumn": 34,
    "endLine": 6,
    "line": 6,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
  {
    "column": 15,
    "endColumn": 36,
    "endLine": 7,
    "fix": {
      "range": [
        225,
        244,
      ],
      "text": "Readonly<ReadonlySet<string>>",
    },
    "line": 7,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
  {
    "column": 15,
    "endColumn": 44,
    "endLine": 8,
    "fix": {
      "range": [
        264,
        291,
      ],
      "text": "Readonly<ReadonlyMap<string, string>>",
    },
    "line": 8,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
]
`;

exports[`prefer-immutable-types > typescript > allows for user suggestions 1`] = `
[
  {
    "column": 15,
    "endColumn": 32,
    "endLine": 1,
    "line": 1,
    "message": "Return type should have an immutability of at least "ReadonlyDeep" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
    "suggestions": [
      {
        "data": {
          "message": "Replace with: ReadonlyDeep<{ foo: string }>",
        },
        "desc": "Replace with: ReadonlyDeep<{ foo: string }>",
        "fix": {
          "range": [
            16,
            31,
          ],
          "text": "ReadonlyDeep<{ foo: string }>",
        },
        "messageId": "userDefined",
      },
    ],
  },
]
`;

exports[`prefer-immutable-types > typescript > reports mutable records return types 1`] = `
[
  {
    "column": 15,
    "endColumn": 32,
    "endLine": 1,
    "line": 1,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
]
`;

exports[`prefer-immutable-types > typescript > reports mutable records return types and suggests a fix for ReadonlyShallow 1`] = `
[
  {
    "column": 15,
    "endColumn": 32,
    "endLine": 1,
    "line": 1,
    "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
    "suggestions": [
      {
        "data": {
          "message": "Surround with Readonly.",
        },
        "desc": "Surround with Readonly.",
        "fix": {
          "range": [
            16,
            31,
          ],
          "text": "Readonly<{ foo: string }>",
        },
        "messageId": "userDefined",
      },
    ],
  },
]
`;

exports[`prefer-immutable-types > typescript > reports non-immutable map return types 1`] = `
[
  {
    "column": 15,
    "endColumn": 44,
    "endLine": 1,
    "line": 1,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
]
`;

exports[`prefer-immutable-types > typescript > reports non-immutable set return types 1`] = `
[
  {
    "column": 15,
    "endColumn": 36,
    "endLine": 1,
    "line": 1,
    "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
  },
]
`;

exports[`prefer-immutable-types > typescript > suggest multiple fixes for collections for ReadonlyShallow 1`] = `
[
  {
    "column": 15,
    "endColumn": 30,
    "endLine": 1,
    "line": 1,
    "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
    "suggestions": [
      {
        "data": {
          "message": "Use ReadonlyArray instead of Array.",
        },
        "desc": "Use ReadonlyArray instead of Array.",
        "fix": {
          "range": [
            16,
            29,
          ],
          "text": "ReadonlyArray<string>",
        },
        "messageId": "userDefined",
      },
      {
        "data": {
          "message": "Surround with Readonly.",
        },
        "desc": "Surround with Readonly.",
        "fix": {
          "range": [
            16,
            29,
          ],
          "text": "Readonly<Array<string>>",
        },
        "messageId": "userDefined",
      },
    ],
  },
  {
    "column": 15,
    "endColumn": 25,
    "endLine": 2,
    "line": 2,
    "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
    "suggestions": [
      {
        "data": {
          "message": "Prepend with readonly.",
        },
        "desc": "Prepend with readonly.",
        "fix": {
          "range": [
            49,
            57,
          ],
          "text": "readonly string[]",
        },
        "messageId": "userDefined",
      },
      {
        "data": {
          "message": "Surround with Readonly.",
        },
        "desc": "Surround with Readonly.",
        "fix": {
          "range": [
            49,
            57,
          ],
          "text": "Readonly<string[]>",
        },
        "messageId": "userDefined",
      },
    ],
  },
  {
    "column": 15,
    "endColumn": 28,
    "endLine": 3,
    "line": 3,
    "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
    "suggestions": [
      {
        "data": {
          "message": "Use ReadonlySet instead of Set.",
        },
        "desc": "Use ReadonlySet instead of Set.",
        "fix": {
          "range": [
            77,
            88,
          ],
          "text": "ReadonlySet<string>",
        },
        "messageId": "userDefined",
      },
      {
        "data": {
          "message": "Surround with Readonly.",
        },
        "desc": "Surround with Readonly.",
        "fix": {
          "range": [
            77,
            88,
          ],
          "text": "Readonly<Set<string>>",
        },
        "messageId": "userDefined",
      },
    ],
  },
  {
    "column": 15,
    "endColumn": 36,
    "endLine": 4,
    "line": 4,
    "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "returnType",
    "nodeType": "TSTypeAnnotation",
    "ruleId": "prefer-immutable-types",
    "severity": 2,
    "suggestions": [
      {
        "data": {
          "message": "Use ReadonlyMap instead of Map.",
        },
        "desc": "Use ReadonlyMap instead of Map.",
        "fix": {
          "range": [
            108,
            127,
          ],
          "text": "ReadonlyMap<string, string>",
        },
        "messageId": "userDefined",
      },
      {
        "data": {
          "message": "Surround with Readonly.",
        },
        "desc": "Surround with Readonly.",
        "fix": {
          "range": [
            108,
            127,
          ],
          "text": "Readonly<Map<string, string>>",
        },
        "messageId": "userDefined",
      },
    ],
  },
]
`;
